MAGN-6241 Regularly publish searchable, easy to read documentation

ramramps opened the pull request:
### Purpose
Link to YouTrack:
[MAGN-6241](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6241) 

The purpose of this PR is to create documentation for Dynamo. The process has to be automated and for every push to master, the documentation has to be updated automatically.  

XmlDocToMarkDown was modified to create markdown files which will be deployed to github by mkdocs.

To accomplish this :

1. A new GitHub repo - DynamoAPI was created on DynamoDS.
2. Personal Access Token was generated for DynamoDS 
3. AppVeyor.yml file was updated to run mkdocs for every commit and publish only for merge
4. BuildScript.ps1 was added to BuildScripts folder , which creates the markdown files for the DLL's passed as parameters
5. PublishScript.ps1 was added to BuildScripts folder, which publishes the document to github.

Here is the documentation url :  http://dynamods.github.io/DynamoAPI/

### Workflow
1. For every commit, Appveyor builds Dynamo. 
2. If the build is successful,  then Appveyor runs the script  BuildScript.ps1. This script runs thexmldoc tool and creates the markdown files.
3. If the commit is on master, then Appveyor runs the script PublishScript.ps1. This script installs MkDocs for that instance and then publish the markdown files to github using mkdocs. 

### Running on Desktop

To test this on Desktop,

1. Install MkDocs from http://mkdocs.org/  For windows add the mkdocs to (environment) PATH variable.
2. There is a test library in XmlDocs solution. In XmlDocsToMarkdown project, change the build properties to point DocumentationTestLibrary. It has to be <path to DocumentationTestLibrary.dll> <path to DocumentationTestLibrary.xml>
3. Create methods, events, properties on the test library. Otherwise, it is difficult to debug and identify the issue. 
4. Run XmlDocToMarkDown. Inside debug (or release), you can see the docs folder created. 
5. Now go to the XmlDocToMarkDown project folder, copy the index.md file and paste it inside docs folder (which is in debug (or release).
6 . Open command prompt and run the command Mkdocs serve. 
7.  Go to localhost:8000 and you can see the documentation.

### Declarations

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files

### Reviewers
@ikeough

ikeough commented on appveyor.yml:
```diff
@@ -4,8 +4,27 @@ platform: Any CPU
+  dynamo_solution: src/Dynamo.All.2013.sln
+  xmlDocs_solution: tools/XmlDocToMarkdown/XmlDocToMarkdown.sln
+  access_token:
+    secure:a53fed0916140fa078b4559a240e740871a0a67f
```
We can't have this access token in the .yml on a public repo. This token has now been revoked. Let's talk about how to make this private.

ikeough commented on tools/XmlDocToMarkdown/XmlDocToMarkdown/Program.cs:
```diff
@@ -79,11 +97,13 @@ private static void GenerateDocYaml()
                 tw.WriteLine("site_name: Dynamo");
-                tw.WriteLine("pages:");
-                tw.WriteLine("- Home: index.md");
+                tw.WriteLine("repo_url: http://ramramps.github.io/DocumentTesting/");
```
These need to be updated for production. And let's move these as strings to the top of the file.

ikeough commented:
A few things that we need to fix:
- We need return types.
- Grids for parameters/descriptions are still showing "------" sometimes.
- Parameter names in grids don't need ":" after them.
- Constructors need to be documented.
- Returns are sometimes being written into the parameters grid.

![image](https://cloud.githubusercontent.com/assets/1139788/10033095/1fa04d1e-613c-11e5-9b37-cf5e21f0a1d5.png)

![image](https://cloud.githubusercontent.com/assets/1139788/10033136/76d43334-613c-11e5-9add-7c3132ffa103.png)

![image](https://cloud.githubusercontent.com/assets/1139788/10033147/92cb4938-613c-11e5-9412-83d22d72cf3c.png)

ikeough commented:
The "Home" page should be cleaned out, with just a link to the DynamoBIM.org site and to the github repo.

![image](https://cloud.githubusercontent.com/assets/1139788/10033209/f7f5669a-613c-11e5-8106-ed45ce57c3ce.png)

ramramps commented:
@ikeough - PTAL. 
1. Added Constructors
2. Updated class Descriptions. (no syle added for now).
3. Added return type for methods
4. removed the weird 1 from the methods
5. updated interface
6. updated table for parameters.

ikeough commented:
Parameter tables are still showing -------
![image](https://cloud.githubusercontent.com/assets/1139788/10107513/87ec6fb6-6370-11e5-93de-428c95f36337.png)

ikeough commented:
Properties now have () which make them look like methods. And, we should documentation as to whether the properties have `get` or `set` accessors.
![image](https://cloud.githubusercontent.com/assets/1139788/10107545/b7d13a9a-6370-11e5-88cd-7f7f1e9b85cc.png)

ikeough commented:
I think we should format methods, properties, and events with code formatting like:
`void AddHomeWorkspace()`

as opposed to 

Void AddHomeWorkspace()

ikeough commented:
When a block is empty, for example when there are no public constructors defined, we should either remove the block, or include some text like:  
> No public constructors defined
or
> No public events defined

ikeough commented:
Some parameter description tables are getting multiple "Name" and "Description" lines:
![image](https://cloud.githubusercontent.com/assets/1139788/10107710/cce3de50-6371-11e5-9793-d2e76ca2e118.png)

ikeough commented:
If the `<returns>` tag is empty for a member, we should put something sensible - like the return type. For example, for `NoteModel.LoadNoteFromXml(XmlNode note)`, we could do:

> Returns 
>`NoteModel`

![image](https://cloud.githubusercontent.com/assets/1139788/10107770/1db258b6-6372-11e5-90bb-27776b33cd86.png)

ikeough commented:
The image in the upper corner is a broken link:
![image](https://cloud.githubusercontent.com/assets/1139788/10107826/735b88f0-6372-11e5-9b31-39b59d54ff48.png)

ikeough commented:
Sometimes the class description is formatted in the same weight as the class name:
![image](https://cloud.githubusercontent.com/assets/1139788/10107855/a06bca3a-6372-11e5-808b-a71c35411c74.png)

ikeough commented:
@ramramps We're getting closer :)

- If we put the stability index in the top bar of the block, we can remove the superscript from the member name.
- Parameter names are now formatting with the title heading ":" is back.
- Can we try using code formatting on the member signatures? It'll probably won't work with the hyperlinks we have to member types, but it will be much more readable as code. I think it's a fair trade.


![image](https://cloud.githubusercontent.com/assets/1139788/10142371/388ccbb0-65c7-11e5-9db1-b1222157d87d.png)
